### PR TITLE
Support for Custom GH Instances + Submodules

### PR DESCRIPTION
**Pull request by the-powerpointer**
Formerly the GithubRepoLoader was hardcoded to support only github.com.
Especially when working with GitHub Enterprise instances, one need to target custom GitHub instances.
This PR provides the support via additional parameters `baseUrl` and `apiUrl`.

Furthermore the GithubRepoLoader formerly broke when it was used with a repository using Git submodules.
This PR also provides the support for recursively processing submodules (as long as they stay within the boundaries of the same GitHub instance).

### CODE REVIEW

File: langchain/src/document_loaders/web/github.ts [hunk 1]
```diff
@@ -249,22 +435,23 @@ export class GithubRepoLoader
...
+          currentDirectoryFilePromises.push(this.fetchFileContentWrapper(file));
+        } catch (e) {
+          this.handleError(`Failed to fetch file content: ${file.path}, ${e}`);
         }
+      } else if (this.recursive) {
+        currentDirectoryDirectoryPromises.push(
+          this.processDirectory(file.path)
+        );
       }
     }
```
[the remark this answers is not in the document]
Comment (the-powerpointer) [reply langchain/src/document_loaders/web/github.ts:457]: the affected code itself was not touched, only rearranged for better readability.

File: langchain/src/document_loaders/web/github.ts [hunk 2]
```diff
@@ -105,14 +135,25 @@ export class GithubRepoLoader
```
[the remark this answers is not in the document]
Comment (the-powerpointer) [reply langchain/src/document_loaders/web/github.ts:135]: The parameter `accessToken` was not touched in this PR, it existed before.

File: langchain/src/document_loaders/web/github.ts [hunk 3]
```diff
@@ -123,16 +164,22 @@ export class GithubRepoLoader
...
+    this.baseUrl = baseUrl;
+    this.apiUrl = apiUrl;
     const { owner, repo, path } = this.extractOwnerAndRepoAndPath(githubUrl);
     this.owner = owner;
     this.repo = repo;
     this.initialPath = path;
     this.branch = branch;
     this.recursive = recursive;
+    // processing submodules without processing contents of other directories makes no sense
+    this.processSubmodules = processSubmodules && recursive;
```
Comment (jacoblee93) [inline langchain/src/document_loaders/web/github.ts:176]: nit: throw here if `processSubmodules` is `true` and `recursive` is `false`?

File: langchain/src/document_loaders/web/github.ts [hunk 4]
```diff
@@ -180,13 +228,147 @@ export class GithubRepoLoader
...
+    } else if (!submoduleInfo.path.startsWith(this.initialPath)) {
+      this.log(
+        `Ignoring submodule ${submoduleInfo.url}, as it is not on initial path.`
+      );
+      return [];
+    } else {
+      this.log(
+        `Accessing submodule ${submoduleInfo.name} (${submoduleInfo.url})...`
+      );
+      return new GithubRepoLoader(submoduleInfo.url, {
```
Comment (jacoblee93) [inline langchain/src/document_loaders/web/github.ts:357]: Clever

**Comment by jacoblee93** [comment]
Thank you very much for this!

**Comment by the-powerpointer** [comment]
@jacoblee93 Thank you for the fast reaction. Nice experience for my first PR to langchainjs 😏

**Pull request merged**